stdlib/InteractiveUtils/src/InteractiveUtils.jl

```diff
@@ -37,24 +37,31 @@ function varinfo(m::Module=Main, pattern::Regex=r""; all::Bool = false, imported
     sortby in (:name, :size, :summary) || throw(ArgumentError("Unrecognized `sortby` value `:$sortby`. Possible options are `:name`, `:size`, and `:summary`"))
     rows = Vector{Any}[]
     workqueue = [(m, ""),]
-    while !isempty(workqueue)
+    pushlock = Base.ReentrantLock()
+    @sync while !isempty(workqueue)
         m2, prep = popfirst!(workqueue)
         for v in names(m2; all, imported)
-            if !isdefined(m2, v) || !occursin(pattern, string(v))
-                continue
-            end
-            value = getfield(m2, v)
-            isbuiltin = value === Base || value === Main || value === Core
-            if recursive && !isbuiltin && isa(value, Module) && value !== m2 && nameof(value) === v && parentmodule(value) === m2
-                push!(workqueue, (value, "$prep$v."))
-            end
-            ssize_str, ssize = if isbuiltin
-                    ("", typemax(Int))
-                else
-                    ss = summarysize(value)
-                    (format_bytes(ss), ss)
+            Threads.@spawn begin
```

Comment:
`popfirst!(workqueue)` has to be protected by the lock too. Alternatively, move `@sync` to `for v in names(m2; all, imported)` (which is probably a bit better because tasks can be GC'ed earlier).

But for tree traversal like this, you don't need (and likely don't want) a lock. I'd write it as a more straightforward recursive function and replace all the recursion with spawn. Then concatenate returned rows after sync so that each task (tree) mutates its own vector.